### ISSUE TITLE
control or delay building indexes

### ISSUE THREAD

**Issue by barbalex** [Nozbe/WatermelonDB#908]
My app works offline. So on first load it downloads all data. Which for some users can be a few 10'000 records.

Downloading usually takes about five to ten seconds. That is o.k. and the app shows the user what data is being downloaded. So o.k. UX.

After the first (initial) load the app only needs to fetch data that has changed since the last start. So not a lot.

The problem is: After the data is initially loaded the app freezes for quite a while. Depending on the device (powerful pc to old weak mobile phone) this will take 5 seconds up until minutes. This happens only on first load, later it is not an issue.

I don't know how to control this and show the user what is going on.

Most likely this happens due to indexedDB building indexes.

Unfortunately there seems to be no way to detect or control this directly in indexedDB: https://stackoverflow.com/a/39212329/712005.

Would there be a way to tell watermelonDB to build indexes without including it in the db schema? I am thinking of adding indexes one by one after the first load has ended. Trying not to block the UI as much. And enabling showing the user exactly what is happening and how it is progressing.

**Comment by radex** [comment]
> Would there be a way to tell watermelonDB to build indexes without including it in the db schema? I am thinking of adding indexes one by one after the first load has ended. 

there are no IDB indexes used other than the primary (key) index, so there isn't anything to skip here.

> Most likely this happens due to indexedDB building indexes.

Do not guess! Measure! Open up Chrome web inspector, open the profiler, and run a profile (in development mode or without a minifier, so that symbol names are not obscured), and it will tell you exactly what is the problem. If you find that troublesome, attach the exported chrome profile here and I may be able to give you tips

**Comment by barbalex** [comment]
@radex 

During some quiet time on the balcony bathing in the sun I suddenly realized that I _did_ have some control: By importing the data one table at a time. Which I tried to achieve by using a `setTimeout` before inserting the data of a table into WatermelonDB: https://github.com/barbalex/vermehrung/blob/master/src/utils/processSubscriptionResult.js#L66

The result was HUGE:
- the ui remains responsive - it now manages to tell the user each table being imported
- lag/freezing has virtually disappeared. I assume that by importing tens of thousands of records at once the app had hit some hard limits that aren't hit when the tables are imported one by one. But as you say: I should rather measure

I also realized that my db schema was not optimal for the offline case: more than 50% of the records downloaded had to do with settings rather than the hard data recorded by users 🙄. With some simple changes I could more than half the number of records that had to be downloaded, from about 50'000 to about 23'000, which now is >95% real user recorded data. This also helped tangibly.

A quick performance run gave this:
<img width="1327" alt="2021-01-11_21h01_08" src="https://user-images.githubusercontent.com/669783/104232166-3e0ba700-5450-11eb-97f5-f4dd21bf0596.png">

Which, I have to be honest, I will have to spend some time breeding over until I understand what is going on. But now that the data loads so fast, that is fortunately not a priority any more. Thanks to your great tool.

Thanks a lot for your help! And of course: for giving us this great tool.
